lib/src/clojure_lsp/feature/diagnostics.clj

```diff
@@ -144,42 +143,59 @@
         (when (not= :unknown (shared/uri->file-type uri))
           (sync-lint-file! uri db))))))
 
-(defn ^:private pmap-light
-  "Call claypoole pmap with less threads than pmap to avoid topping cpu."
-  [f coll]
-  (let [threadpool-size (int (Math/ceil (/ (.. Runtime getRuntime availableProcessors) 3)))]
-    (cp/upmap threadpool-size f coll)))
-
 (defn ^:private unused-public-vars-lint!
-  [definitions project-analysis {:keys [config reg-finding!]} max-parallelize? db]
-  (let [parallelize-fn (if max-parallelize? pmap pmap-light)]
-    (->> definitions
-         (remove (partial exclude-public-diagnostic-definition? config))
-         (parallelize-fn #(when (= 0 (count (q/find-references project-analysis % false db)))
-                            %))
-         (remove nil?)
-         (mapv #(reg-unused-public-var-element! % reg-finding! config))
-         (group-by :filename))))
-
-(defn ^:private project-definitions [project-analysis]
-  (concat (q/find-all-var-definitions project-analysis)
-          (q/find-all-keyword-definitions project-analysis)))
-
-(defn ^:private file-definitions [project-analysis filename]
-  (concat (q/find-var-definitions project-analysis filename false)
-          (q/find-keyword-definitions project-analysis filename)))
+  [var-defs kw-defs project-analysis {:keys [config reg-finding!]}]
+  (let [var-definitions (remove (partial exclude-public-diagnostic-definition? config) var-defs)
+        var-nses (set (map :ns var-definitions)) ;; optimization to limit usages to internal namespaces, or in the case of a single file, to its namespaces
+        var-usages (into #{}
+                         (comp
+                           (q/xf-all-var-usages-to-namespaces var-nses)
+                           (map (juxt :to :name)))
```

Comment:
Maybe extract the juxt just like you did for `kw-signature`? I don't think though it would have a significant performance improvement

Reply:
`kw-signature` is the same for both definitions and usages. But for vars, it's `(juxt :ns :name)` for definitions and `(juxt :to :name)` for usages. I'm not sure... I'll try `var-usage-signature` and `var-def-signature`, but might also inline `kw-signature`.